python-package/lightgbm/dask.py

```diff
@@ -197,29 +196,20 @@ def _train(client, data, label, params, model_factory, sample_weight=None, group
     """
     params = deepcopy(params)
 
-    # Split arrays/dataframes into parts. Arrange parts into tuples to enforce co-locality
+    # Split arrays/dataframes into parts. Arrange parts into dicts to enforce co-locality
     data_parts = _split_to_parts(data=data, is_matrix=True)
     label_parts = _split_to_parts(data=label, is_matrix=False)
+    parts = [{'data': x, 'label': y} for (x, y) in zip(data_parts, label_parts)]
 
     if sample_weight is not None:
         weight_parts = _split_to_parts(data=sample_weight, is_matrix=False)
-    else:
-        weight_parts = None
+        for i, d in enumerate(parts):
+            parts[i] = {**d, 'weight': weight_parts[i]}
```

Comment:
~Will this result in each of the existing parts (data, label) in `d` being copied when they're passed back in? I think it might.~ Ok ignore that, I tested and was wrong about how `**` worked.

Still, I have a slight preference just modifying `parts[i]` directory, taking advantage of the fact that we know it's already a dictionary. 

I think this is a little easier to read, what do you think?

```suggestion
        for i in len(parts):
            parts[i]['weight'] = weight_parts[i]
```

Reply:
I absolutely agree with James! But it should be `for i in range(len(parts)):`

Reply:
ha! typed too fast, thanks. Edited it.

Reply:
yep yeah, this is easier to understand, all for it

Reply:
Addressed in 0d36cc4